p2p/net/swarm/black_hole_detector.go

```diff
@@ -0,0 +1,208 @@
+)
+
+type outcome int
+
+const (
+	outcomeSuccess outcome = iota
+	outcomeFailed
+)
+
+type blackHoleState int
+
+const (
+	blackHoleStateAllowed blackHoleState = iota
+	blackHoleStateBlocked
+)
+
+// blackHoleFilter provides black hole filtering logic for dials. On detecting a black holed
+// network environment, subsequent dials are blocked and only 1 dial every n requests is allowed.
+// This should be used in conjunction with an UDP or IPv6 address filter to detect UDP or
+// IPv6 black hole.
+// Requests are blocked if the success fraction in the last n outcomes is less than
+// minSuccessFraction. If a request succeeds in Blocked state, the filter state is reset and n
+// subsequent requests are allowed before reevaluating black hole status. Evaluating over n
+// outcomes avoids situations where a dial was cancelled because a competing dial succeeded,
+// the address was unreachable, and other false negatives.
+type blackHoleFilter struct {
+	// n is the minimum number of completed dials required before we start blocking.
+	// Every nth request is allowed irrespective of the status of the detector.
+	n int
+	// minSuccessFraction is the minimum success fraction required to allow dials.
+	minSuccessFraction float64
+	// name for the detector.
+	name string
+
+	// requests counts number of dial requests up to n. Resets to 0 every nth request.
+	requests int
+	// outcomes of the last `n` allowed dials
+	outcomes []outcome
+	// successes is the count of successful dials in outcomes
+	successes int
+	// failures is the count of failed dials in outcomes
+	failures int
+	// state is the current state of the detector
+	state blackHoleState
+
+	mu            sync.Mutex
+	metricsTracer MetricsTracer
+}
+
+// RecordOutcome records the outcome of a dial. A successful dial will change the state
+// of the filter to Allowed. A failed dial only blocks subsequent requests if the success
+// fraction over the last n outcomes is less than the minSuccessFraction of the filter.
+func (b *blackHoleFilter) RecordOutcome(success bool) {
+	if b == nil {
+		return
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.state == blackHoleStateBlocked && success {
+		// If the call succeeds in a blocked state we reset to allowed.
+		// This is better than slowly accumulating values till we cross the minSuccessFraction
+		// threshold since a blackhole is a binary property.
+		b.reset()
+		return
+	}
+
+	if success {
+		b.successes++
+		b.outcomes = append(b.outcomes, outcomeSuccess)
+	} else {
+		b.failures++
+		b.outcomes = append(b.outcomes, outcomeFailed)
+	}
+
+	if len(b.outcomes) > b.n {
+		if b.outcomes[0] == outcomeSuccess {
+			b.successes--
+		} else {
+			b.failures--
+		}
+		b.outcomes = b.outcomes[1 : b.n+1]
+	}
+
+	b.updateState()
+	b.trackMetrics()
+}
+
+func (b *blackHoleFilter) IsAllowed() (state blackHoleState, isAllowed bool) {
+	if b == nil {
+		return blackHoleStateAllowed, true
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.requests++
```

Comment:
Why are you incrementing the request counter here? Just the fact that a dial is allowed doesn't mean that it actually happened. It might have been delayed by the smart dialing logic for example.

Reply:
This is why I moved the IsAllowed check to [here](https://github.com/sukunrt/go-libp2p/blob/blackhole-detector/p2p/net/swarm/swarm_dial.go#L399-L406) See the comment. 
If I have this check with filtering we run in to the issue you mentioned here.

Reply:
We need to increment the counter here. There is no other way to handle this. If we are blocking request there will be no call to RecordOutcome, so this counter needs to be reset in this method.